core/src/main/java/io/grpc/internal/AbstractServerImplBuilder.java

```diff
@@ -228,13 +242,26 @@ public Server build() {
     ArrayList<ServerStreamTracer.Factory> tracerFactories =
         new ArrayList<ServerStreamTracer.Factory>();
     if (statsEnabled) {
-      StatsContextFactory statsFactory =
-          this.statsFactory != null ? this.statsFactory : Stats.getStatsContextFactory();
-      if (statsFactory != null) {
-        CensusStatsModule censusStats =
-            new CensusStatsModule(statsFactory, GrpcUtil.STOPWATCH_SUPPLIER, true, recordStats);
-        tracerFactories.add(censusStats.getServerTracerFactory());
-      }
+      Tagger tagger = this.tagger != null ? this.tagger : Tags.getTagger();
+      TagContextBinarySerializer tagCtxSerializer =
+          this.tagCtxSerializer != null
+              ? this.tagCtxSerializer
+              : Tags.getTagPropagationComponent().getBinarySerializer();
+      StatsRecorder statsRecorder =
+          this.statsRecorder != null ? this.statsRecorder : Stats.getStatsRecorder();
+      // // TODO: How do we check whether stats is enabled, now that the StatsRecorder is always
+      // // non-null? Uncommenting this line causes test failures.
```

Comment:
Most tests pass in the fake implementations from `StatsTestUtils`, but `Stats.getState()` returns `DISABLED`. If you don't initialize `CensusStatsModule`, these tests will fail.

A possibly better approach, is to make `CensusStatsModule` public, and change the three-argument Census impl setter of the builder to `overrideCensusStatsModule(CensusStatsModule)`. Tests will create `CensusStatsModule` out of the impls from `StatsTestUtils`, and here you can do:
```java
if (statsEnabled) {
  CensusStatsModule censusStats = this.censusStatsOverride;
  if (censusStats == null && Stats.getStats() == StatsCollectionState.ENABLED) {
    CensusStatsModule censusStats =
        new CensusStatsModule(...);
  }
  if (censusStats != null) {
    tracerFactories.add(censusStats.getServerTracerFactory());
  }
}
```

Reply:
@zhangkun83 I refactored the code for overriding the OpenCensus implementation.  There are two parts that I'm not sure about.

- CensusStatsModule no longer caches the StatsClientInterceptor and ServerTracerFactory.  It creates a new object whenever getServerTracerFactory or getClientInterceptor is called.  Is that correct?
- I'm not completely sure I used the right values for `recordStats`, especially in CensusModulesTest.